### PR TITLE
Refactoring: replace `Url`-based connection strings with a hand-crafted struct

### PR DESCRIPTION
**Pull request by yeputons**
Previous attempt: #2730

My thoughts are in the first commit's description.

### CODE REVIEW

File: libs/utils/src/postgres_connection.rs
Comment (yeputons) [inline libs/utils/src/postgres_connection.rs:5]: There is a similar function [in proxy](https://github.com/neondatabase/neon/blob/f564dff0e320a859ab9eb8b40caa9a70599c829a/proxy/src/auth/backend/console.rs#L252-L255), but I wouldn't touch it in this PR, so there's less to review. Maybe a separate PR, one can argue that IPv6 support is good.

File: control_plane/src/pageserver.rs
Comment (yeputons) [inline control_plane/src/pageserver.rs:99]: `no_user`/`no_db` were never parsed into `PgConnectionConfig` and are ok to ignore.

File: libs/utils/src/postgres_connection.rs
Comment (vadim2404) [inline libs/utils/src/postgres_connection.rs:129]: Won't be easier to put "******" here directly instead of creating a temporary struct for one field debugging?

Comment (yeputons) [reply libs/utils/src/postgres_connection.rs:129]: It will be displayed as `"******"`, which looks exactly like a string of asterisks. Not that it's _that_ confusing, but still.

File: libs/utils/src/postgres_connection.rs
Comment (funbringer) [reply libs/utils/src/postgres_connection.rs:5]: You might want to use [`SocketAddr`](https://doc.rust-lang.org/std/net/enum.SocketAddr.html) or [`ToSocketAddrs`](https://doc.rust-lang.org/std/net/trait.ToSocketAddrs.html) here instead.

In fact, I'll likely use one of those in the proxy.

File: libs/utils/src/lib.rs
Comment (funbringer) [inline libs/utils/src/lib.rs:15]: This module is big enough (200+ lines including tests), so I think it deserves to be a crate in `libs/`. That way you won't re-add `tokio_postgres` to already bloated `utils`.

File: libs/utils/src/postgres_connection.rs
Comment (funbringer) [inline libs/utils/src/postgres_connection.rs:124]: I don't think this is needed. `Option<&str>` implements Debug.

File: libs/utils/src/postgres_connection.rs
Comment (funbringer) [inline libs/utils/src/postgres_connection.rs:53]: PgConnectionConfig -> Self

File: libs/utils/src/postgres_connection.rs
Comment (funbringer) [inline libs/utils/src/postgres_connection.rs:90]: Most of the time rust's naming conventions use `to_` (ref to owned), `into_` (owned to owned) and `as_` (ref to ref), so this should be `to_tokio_postgres_config`.

File: libs/utils/src/postgres_connection.rs
Comment (funbringer) [inline libs/utils/src/postgres_connection.rs:68]: Why allow changing everything but this? What's special about host and port?

File: libs/utils/src/postgres_connection.rs
Comment (funbringer) [inline libs/utils/src/postgres_connection.rs:103]: What's going on here?

File: libs/utils/src/postgres_connection.rs
Comment (funbringer) [inline libs/utils/src/postgres_connection.rs:78]: I guess you could use `self.options.extend()` here.

**Review by funbringer — COMMENTED** [review]
Thanks for the hard work! I've left a bunch of comments.

File: libs/utils/src/lib.rs
Comment (yeputons) [reply libs/utils/src/lib.rs:15]: Moved into a separate `postgres_connection` crate

File: libs/utils/src/postgres_connection.rs
Comment (yeputons) [reply libs/utils/src/postgres_connection.rs:124]: Added a comment. Let me know if you still think it's better to display `"REDACTED"` instead of `REDACTED` to save some lines, your call. That's the way `tokio_postgres::Config` does it and it distinguishes between "a redacted string" and "a string which contains the value `REDACTED`".

File: libs/utils/src/postgres_connection.rs
Comment (yeputons) [reply libs/utils/src/postgres_connection.rs:103]: Added a comment.

File: libs/utils/src/postgres_connection.rs
Comment (yeputons) [reply libs/utils/src/postgres_connection.rs:78]: Good idea, thanks. Also renamed to `extend_options`.

File: libs/utils/src/postgres_connection.rs
Comment (yeputons) [reply libs/utils/src/postgres_connection.rs]: `Host` also includes unresolved hostnames. I suspect it's more obvious to resolve on connect rather than on configuration parsing. So I'd keep it as it is.

**Comment by yeputons** [comment]
@funbringer I've just addressed all the comments, marked ones I consider unambiguous as resolved, two unresolved are rest. I did not change any other code, rebase was without conflicts.

File: libs/postgres_connection/src/lib.rs
Comment (vadim2404) [inline libs/postgres_connection/src/lib.rs:163]: Do you really need to have the dev-dependency for this line? 

maybe you can replace it with a short method in this module, can't you?

```rust
fn stub_host() -> Host  {
    Host::Domain("stub.host.example".to_owned())
}
```

File: libs/utils/src/postgres_connection.rs
Comment (funbringer) [reply libs/utils/src/postgres_connection.rs]: I see! How about this:

```rust
struct Kek {
    value: u32,
}

impl std::fmt::Debug for Kek {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Kek").field("value", &format_args!("REDACTED")).finish()
    }
}

fn main() {
    println!("{:?}", Kek { value: 10 });
}
```

I find `format_args!` immensely useful, because it's essentially `format!` but w/o allocations.

File: libs/postgres_connection/src/lib.rs
Comment (funbringer) [reply libs/postgres_connection/src/lib.rs:163]: I have something different on my mind. [`Host`](https://docs.rs/url/latest/url/enum.Host.html) can be parameterized by `&str` instead of String, so can make `new_host_port` more generic and  call [`to_owned`](https://docs.rs/url/latest/url/enum.Host.html#method.to_owned) inside.

This will let us use

```rust
const STUB_HOST: Host<&str> = Host::Domain("stub.host.example");
```

API users will likely benefit from this too.

Comment (vadim2404) [reply libs/postgres_connection/src/lib.rs:163]: More elegant proposal than mine! 

My idea was devoted not to adding new dev-dependency for this one line, but

Comment (yeputons) [reply libs/postgres_connection/src/lib.rs:185]: Nice idea! Unfortunately, I bumped into the comparison code: `Host<String>` can only be compared with `Host<String>`, not `Host<&str>`. Therefore I have to convert `STUB_HOST: Host<&str>` to `Host<String>` in `assert_eq!` anyway. Even if I support both `Host<String>` and `Host<&str>` in `new_host_port` (the only way to "overload" in Rust is to create a trait, isn't it..?), I still get `.to_owned()` in the test.

So I prefer to leave the code as is. What I have here is a global const, initialized once, not a value generated dynamically on the fly for each test (e.g. `assert_eq!` do not clone the const), so `Lazy` expresses the intention better. Also, we use `once_cell` heavily in our code a lot already, and its standard counterpart is in progress (https://github.com/rust-lang/rust/issues/74465).

File: libs/postgres_connection/src/lib.rs
Comment (yeputons) [reply libs/postgres_connection/src/lib.rs]: Good idea, I forgot about splitting from the right. Splitting from the left does not work with `[::1]`, so I discarded it altogether.

File: libs/postgres_connection/src/lib.rs
Comment (yeputons) [reply libs/postgres_connection/src/lib.rs]: Added `Cow`, turned out to be not much code. However, there are still some allocations in the worst case (each option contains a char to be escaped). Not a problem, I think, this function is called infrequently.

If we strictly aim for fewer allocations, I prefer something imperative which guarantees O(1) allocations (depending on `String`'s growth factor, if it's 2, we have 2 allocations at most):

```rust
            let mut options_str = String::new();
            options_str.reserve(
                self.options.iter().map(|s| s.len()).sum::<usize>()
                    + std::cmp::max(1, self.options.len())
                    - 1,
            );
            let mut first_option = true;
            for opt in self.options.iter() {
                if !first_option {
                    options_str.push(' ');
                }
                first_option = false;
                for c in opt.chars() {
                    match c {
                        '\\' => options_str.push_str("\\\\"),
                        ' ' => options_str.push_str("\\ "),
                        c => options_str.push(c),
                    }
                }
            }
            config.options(&options_str);
```

Rust's `std::str::escape_debug` [does something with custom iterators](https://doc.rust-lang.org/src/core/str/mod.rs.html#2473), so I suspect there is no faster standard way. I'm wondering if there is a functional way to write the escaping logic. I've tried like this, but I need an iterator whose `Item` alternates between `Map` and `Once`, so it doesn't compile:

```rust
                &self
                    .options
                    .iter()
                    // Iter<Iter<&str>>, options, each represented as a iterator of chunks
                    .map(|s| {
                        s.split("").map(|c| match c {
                            "\\" => "\\\\",
                            " " => "\\ ",
                            c => c,
                        })
                    })
                    .intersperse(std::iter::once(" ")) // TODO: use impl from std once it's stabilized
                    .flatten()
                    .collect::<String>(),
```

File: libs/postgres_connection/src/lib.rs
Comment (funbringer) [reply libs/postgres_connection/src/lib.rs]: Unfolding since I still have something to add here.

```rust
>> "-0".parse::<u16>() // can't reproduce
Err(ParseIntError { kind: InvalidDigit })
>> "+10".parse::<u16>() // this is unfortunate, yes
Ok(10)
```

I think we could add an `is_numeric` check on top if you'd like to.

File: libs/postgres_connection/src/lib.rs
Comment (funbringer) [reply libs/postgres_connection/src/lib.rs:185]: Ok! It's definitely a nice-to-have, not a must. I might revisit this later on my own.

File: libs/postgres_connection/src/lib.rs
Comment (yeputons) [reply libs/postgres_connection/src/lib.rs]: Good catch, thank you. I've missed that `if` while reading `from_str_radix`.

**Review by funbringer — APPROVED** [review]
Cool! This is a good start. We'll see how this changes in time.

File: pageserver/src/walreceiver/connection_manager.rs
Comment (LizardWizzard) [inline pageserver/src/walreceiver/connection_manager.rs:594]: A bit inconsistent naming, variable remains to be named `new_wal_source_connstr`

File: pageserver/src/walreceiver/connection_manager.rs
Comment (LizardWizzard) [inline pageserver/src/walreceiver/connection_manager.rs:849]: There is no need to use a  `vec!` here. Slice or an array (`&[` or `[`) should be ok
```suggestion
        PgConnectionConfig::new_host_port(host, port).extend_options([
```

File: pageserver/src/walreceiver/connection_manager.rs
Comment (LizardWizzard) [inline pageserver/src/walreceiver/connection_manager.rs:844]: `wal_stream_connection_string` should be probably renamed to `wal_stream_connection_config`

**Review by LizardWizzard — APPROVED** [review]
Thanks! Looks good to me. Left minor comments

Not much to add after @funbringer's review :)

**Comment by yeputons** [comment]
Squashed and rebased, minor conflicts during squash only (I had to rearrange some commits), will merge once tests pass.

**Pull request merged**
